2015/presentations/import_deep_dive/text.md

```diff
@@ -0,0 +1,297 @@
+it makes a *namespace package*: an empty module whose `__path__` contains all
+the directories remembered.
+
+But why do that?
+
+Namespace packages are usually used to implement plugin systems.
+Suppose there is an extensible sound editor, where individual filters
+(like "echo", "reverb" and "speedup") are implemented as Python modules,
+and can be installed individually.
+
+"Individually" means "to different locations".
+Your `sys.path` will usually include several kinds of locations. On a Unixy
+machine, the path will typically have:
+
+* Directories from `$PYTHONPATH`, and the current directory
+* The Python standard library (e.g. `/usr/lib/python3.4`)
+* User-specific locations (such as `~/.local/lib/python3.4/site-packages`)
+* A place for packages installed system-wide (like `/usr/lib/python3.4/site-packages`)
+
+Suppose you install the editor and a built-in "echo" filter system-wide,
+a private "reverb" filter to your home directory, and you have "speedup"
+(which you're currently developing) in the current directory (on `$PYTHONPATH`).
+The paths to these modules would be:
+
+* /usr/lib/python3.4/site-packages/sound_editor_filters/echo.py
+* ~/.local/lib/python3.4/site-packages/sound_editor_filters/reverb.py
+* ./sound_editor_filters/speedup.py
+
+And that repeated `sound_editor_filters` directory becomes a namespace package:
+
+    >>> import sound_editor_filters
+    >>> sound_editor_filters.__path__
+    _NamespacePath([
+        './sound_editor_filters',
+        '~/.local/lib/python3.4/site-packages/sound_editor_filters',
+        '/usr/lib/python3.4/site-packages/sound_editor_filters'])
+
+Importing `sound_editor_filters.echo` will get you the correct module,
+because submodules are searched for in the `__path__` of their parent.
+
+Remember that adding an `__init__.py` to *any* of the `sound_editor_filters`
+directories will cancel the magic. Any directory with `__init__` is a regular
+package, its `__init__`-less siblings are ignored.
+
+Long before namespace packages were added to core Python, `setuptools` contained
+[an utility](https://pythonhosted.org/setuptools/setuptools.html#namespace-packages)
+to declare namespaces manually and explicitly.
+Of course, this still works – and if you need to support Python 2,
+it's the way to go.
+
+
+## Customization examples
+
+Another topic that didn't make it to my talk is customization:
+how to extend importlib to do new, exciting things?
+
+If you're interested in this, I'll just direct you to
+[importlib documentation](https://docs.python.org/3/library/importlib.html).
+Now that you know the terminology and and the "normal" way of how things are
+done, those docs should make much more sense.
+
+Instead, I'll mention some real-world projects that use custom import hooks,
+which you can use as inspiration about what's possible.
+
+Hy (hy.readthedocs.org) allows you to import modules written in a dialect of
+Lisp. Or as its docs put it, it's "lisp-stick on a Python".
```

Comment:
This is a pun on [lipstick on a pig](https://en.wikipedia.org/wiki/Lipstick_on_a_pig), i.e. putting some amount of red stuff on a pig. "Lipstick" is uncountable here.